module/os/windows/spl/spl-kmem.c

```diff
@@ -0,0 +1,6628 @@
+static int kmem_dump_log_idx;
+
+#define	KDI_LOG(cp, stat) {						\
+kmem_dump_log_t *kdl;						\
+if ((kdl = (kmem_dump_log_t *)((cp)->cache_dumplog)) != NULL) {	\
+kdl->stat++;						\
+} else if (kmem_dump_log_idx < KMEM_DUMP_LOGS) {		\
+kdl = &kmem_dump_log[kmem_dump_log_idx++];		\
+kdl->stat++;						\
+kdl->kdl_cache = (cp);					\
+(cp)->cache_dumplog = kdl;				\
+}								\
+}
+
+/* set non zero for full report */
+uint_t kmem_dump_verbose = 0;
+
+/* stats for overize heap */
+uint_t kmem_dump_oversize_allocs = 0;
+uint_t kmem_dump_oversize_max = 0;
+
+static void
+kmem_dumppr(char **pp, char *e, const char *format, ...)
+{
+	char *p = *pp;
+
+	if (p < e) {
+		int n;
+		va_list ap;
+
+		va_start(ap, format);
+		n = vsnprintf(p, e - p, format, ap);
+		va_end(ap);
+		*pp = p + n;
+	}
+}
+
+/*
+ * Called when dumpadm(1M) configures dump parameters.
+ */
+void
+kmem_dump_init(size_t size)
+{
+	if (kmem_dump_start != NULL)
+		zfs_kmem_free(kmem_dump_start, kmem_dump_size);
+
+	if (kmem_dump_log == NULL)
+		kmem_dump_log =
+		    (kmem_dump_log_t *)zfs_kmem_zalloc(
+		    KMEM_DUMP_LOGS * sizeof (kmem_dump_log_t), KM_SLEEP);
+
+	kmem_dump_start = zfs_kmem_alloc(size, KM_SLEEP);
+
+	if (kmem_dump_start != NULL) {
+		kmem_dump_size = size;
+		kmem_dump_curr = kmem_dump_start;
+		kmem_dump_end = (void *)((char *)kmem_dump_start + size);
+		copy_pattern(KMEM_UNINITIALIZED_PATTERN, kmem_dump_start, size);
+	} else {
+		kmem_dump_size = 0;
+		kmem_dump_curr = NULL;
+		kmem_dump_end = NULL;
+	}
+}
+
+/*
+ * Set flag for each kmem_cache_t if is safe to use alternate dump
+ * memory. Called just before panic crash dump starts. Set the flag
+ * for the calling CPU.
+ */
+void
+kmem_dump_begin(void)
+{
+	if (kmem_dump_start != NULL) {
+		kmem_cache_t *cp;
+
+		for (cp = list_head(&kmem_caches); cp != NULL;
+		    cp = list_next(&kmem_caches, cp)) {
+			kmem_cpu_cache_t *ccp = KMEM_CPU_CACHE(cp);
+
+			if (cp->cache_arena->vm_cflags & VMC_DUMPSAFE) {
+				cp->cache_flags |= KMF_DUMPDIVERT;
+				ccp->cc_flags |= KMF_DUMPDIVERT;
+				ccp->cc_dump_rounds = ccp->cc_rounds;
+				ccp->cc_dump_prounds = ccp->cc_prounds;
+				ccp->cc_rounds = ccp->cc_prounds = -1;
+			} else {
+				cp->cache_flags |= KMF_DUMPUNSAFE;
+				ccp->cc_flags |= KMF_DUMPUNSAFE;
+			}
+		}
+	}
+}
+
+/*
+ * finished dump intercept
+ * print any warnings on the console
+ * return verbose information to dumpsys() in the given buffer
+ */
+size_t
+kmem_dump_finish(char *buf, size_t size)
+{
+	int kdi_idx;
+	int kdi_end = kmem_dump_log_idx;
+	int percent = 0;
+	int header = 0;
+	int warn = 0;
+	size_t used;
+	kmem_cache_t *cp;
+	kmem_dump_log_t *kdl;
+	char *e = buf + size;
+	char *p = buf;
+
+	if (kmem_dump_size == 0 || kmem_dump_verbose == 0)
+		return (0);
+
+	used = (char *)kmem_dump_curr - (char *)kmem_dump_start;
+	percent = (used * 100) / kmem_dump_size;
+
+	kmem_dumppr(&p, e, "%% heap used,%d\n", percent);
+	kmem_dumppr(&p, e, "used bytes,%ld\n", used);
```

Comment:
size_t should change size based on the machine word size, but I am not aware of a printf specifier that does on Windows. We probably will need this in a header somewhere:
```
#ifdef _ILP32
#define PRI_SIZE "%lu"
#define PRI_SSIZE "%ld"
#else
#define PRI_SIZE "%llu"
#define PRI_SSIZE "%lld"
#endif
```

Then we could make this:

> kmem_dumppr(&p, e, "used bytes," PRI_SIZE "\n", used);

Reply:
This will cause incorrect numbers to be printed whenever bit 32 or greater is used.